Updated fast calo cluster and new calo trigger algo

bechenard opened the pull request:
A compact yet elegant fast clustering algorithm for all your clustering needs, game night or formal events. It also include a new calorimeter filter - needs to be added to the prolog though.

FNALbuild commented:
Hi @bechenard,
You have proposed changes to files in these packages:
- CaloCluster
- CaloFilters
- CaloDiag

which require these tests: build.

@Mu2e/fnalbuild-users, @Mu2e/write have access to CI actions on [master](/Mu2e/Offline/tree/master).


:hourglass: The following tests have been triggered for d220e20209d42ef4da12e969c9f7b2e0ce9c6285: build  (Build queue is empty)


[About FNALbuild.](https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild) [Code review on Mu2e/Offline.](https://mu2ewiki.fnal.gov/wiki/GitHubWorkflow#Code_Review)

FNALbuild commented:
:sunny: The tests passed at ref d220e20209d42ef4da12e969c9f7b2e0ce9c6285.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged d220e20209d42ef4da12e969c9f7b2e0ce9c6285 at be905d45e6a44c9d40f21aa5d8bde014e3e1713e |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/scons.log). Build time: 13 min 19 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/ceSimReco.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/cosmicSimReco.log) |
| rootOverlaps | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/rootOverlaps.log) |
| g4surfaceCheck | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/g4surfaceCheck.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/g4test_03MT.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 10 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [3 errors 5 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//474/console).
Build artefacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.


Logfiles may also be accessed [at this link.](https://gist.github.com/52d477c143523df37b6072077b6c8a92)

gianipez commented on CaloCluster/fcl/prolog_trigger.fcl:
```diff
@@ -24,43 +24,22 @@ CaloTrigger : {
 {
    producers : {
-       CaloClusterOnline   : @local::CaloClusterOnline
+       CaloClusterFast     : @local::CaloClusterFast
```
this is already defined two lines after

bechenard replied:
Fixed

gianipez commented on CaloCluster/fcl/prolog.fcl:
```diff
@@ -14,7 +14,17 @@ END_PROLOG
 
 
-
+CaloClusterFast:
```
can't this just live in the prolog_trigger file? currently, you have the same blocks defined on both files.

bechenard replied:
Ok, will see what I can do

bechenard replied:
Done

gianipez commented on CaloDiag/test/caloNNDiag.fcl:
```diff
@@ -0,0 +1,74 @@
+      @table::CaloClusterMC.producers	
+
+
+      CaloClusterMaker:
```
can you include the table from the CaloCluster/fcl/prolog_trigger.fcl ?

gianipez commented on CaloFilters/src/FilterEcalNNTrigger_module.cc:
```diff
@@ -0,0 +1,127 @@
+	  const auto& nneighborsId  = cal.crystal(hits[0]->crystalID()).nextNeighbors();
+	            
+	  double e9(hits[0]->energyDep()),e25(hits[0]->energyDep());
+          for (auto hit : hits)
```
can you check the indentation of the whole file?

bechenard replied:
Aargh, pesky editor. It's still this old issue with tab and space. Will fix it

gianipez commented on CaloFilters/src/FilterEcalNNTrigger_module.cc:
```diff
@@ -0,0 +1,127 @@
+
+	  select = true;
+	  //size_t index = std::distance(caloClusters.begin(),clusterIt);
+	  //triginfo->_caloClusters.push_back(art::Ptr<CaloCluster>(caloClustersHandle,index));
```
we need to store the ptrs of all the clusters passing the trigger to ensure the trigger matching

bechenard replied:
I wasn't sure about this. Is it fine if I just save the pointers to this structure, or do I need to add another field?

gianipez replied:
you can have a look at how it's done here: https://github.com/Mu2e/Offline/blob/master/CaloFilters/src/CaloCosmicCalib_module.cc#L221

bechenard replied:
I'm a little confused because my code is already looping over all clusters and accept an event if any one of them passes the trigger.

gianipez commented on CaloFilters/src/FilterEcalNNTrigger_module.cc:
```diff
@@ -0,0 +1,127 @@
+	  //triginfo->_caloClusters.push_back(art::Ptr<CaloCluster>(caloClustersHandle,index));
+     }
+     
+     return select;
```
the logic should be changed to allow making the trigger-matching of all the clusters passing the cut.

bechenard replied:
What do you mean?

gianipez replied:
you can have a look at how it's done here: https://github.com/Mu2e/Offline/blob/master/CaloFilters/src/CaloCosmicCalib_module.cc#L221

bechenard replied:
See your other comment

gianipez reviewed (changes requested):
Hi Bertrand, I made a few comments that need small changes. I also noticed that you forgot to commit the changes needed in `CaloFilters/fcl/prolog_trigger.fcl` to have your new filter instantiated with also its trigger sequence. 

Cheers,

Giani

FNALbuild commented:
:memo: The latest commit by @echenard is timestamped 0:13:45.311325 (hh:mm:ss) in the future. Please check that the date and time is set correctly when creating new commits.

bechenard commented:
Hi Giani,

I fixed the fcl prologs, check if you like them. 
For the filters, I had to guess the values of the trigger bits to merge, I used: TriggerFlag::caloDigis. I also added something in the prolog with a crappy names, so tell me what you would like.
Thanks

FNALbuild commented:
:memo: The latest commit by @echenard is timestamped 0:13:53.268173 (hh:mm:ss) in the future. Please check that the date and time is set correctly when creating new commits.

gianipez commented on CaloFilters/src/FilterEcalNNTrigger_module.cc:
```diff
@@ -0,0 +1,137 @@
+     
+     if (select)
+     {
+        trigInfo._triggerBits.merge(TriggerFlag::caloDigis);
```
`caloDigis` is used by the filters that do hits counting. I think here you should use `caloCluster` instead; this makes a lot of sense because you store the artPtr of the calorimeter cluster.

bechenard replied:
Fixed

gianipez commented on CaloCluster/fcl/prolog_trigger.fcl:
```diff
@@ -26,47 +38,13 @@ CaloTrigger : {
-}  
-
-
+   producers : {CaloClusterFast : @local::CaloClusterFast}   
```
I think that here something confusing happened. `CaloClusterFast` is the module label you defined at the very top. The various module instances should be collected under the `CaloClusterTrigger` namespace.

bechenard replied:
I think I fixed it, please check

gianipez commented on CaloFilters/fcl/prolog_trigger.fcl:
```diff
@@ -118,6 +118,15 @@ CaloFilters : {
+        minMVAScore            : 0.95
+        trigPath               : "CaloNewClusterTrigger"
+        diagLevel              : 0
+    }
```
at the bottom of this file, you should define the sequence to run this filter. This means that you need to instantiate also its dedicated event-prescaler module.

bechenard replied:
I think I did this correctly. If not, I'll let you takeover and do the right thing.

gianipez commented:
Hi @bechenard , I inserted another round of comments. They all should be minor things.

brownd1978 reviewed (approved):
Looks fine to me.

gianipez reviewed (changes requested):
Hi @bechenard ,
I made a few last comments. I also realized that we need to provide a sequence to make a `CaloHits` Collection needed to the `CaloClusterFast` module in the same way we plan to make it in the Online reco, and update the various trigger paths that are suing `CaloClusterFast`.
I think that the CaloRecoDigiMaker should have the possibility to use the `rawProcessor` that should do the job. 
Without this change, all the track trigger paths currently defined will just not work in the Offline environment.

brownd1978 commented on CaloCluster/src/CaloClusterFast_module.cc:
```diff
@@ -1,322 +1,197 @@
+          //find the range around the seed time to search for other hits to form clusters
+          auto iterStart(iterSeed), iterStop(iterSeed);
+          while (iterStop  != hits.end()   && (*iterStop==-1  || caloHits[*iterStop].time() - timeStart < deltaTime_))  ++iterStop;
+          while (iterStart != hits.begin() && (*iterStart==-1 || timeStart - caloHits[*iterStart].time() < deltaTime_)) --iterStart;
```
why not use rbegin,rend instead of rolling your own reverse iterator?

brownd1978 replied:
This is actually a bug if hits has exactly 1 entry.

bechenard replied:
Because you scan back from iterSeed (not the beginning or end of the vector) until you either reach the beginning of the vector or the time difference with the seed hit is too large. This define the lower boundary to include crystal hits in the clustering algorithm (obviously same argument for iterStop going forward). 

if you have a single entry, then iterStart = hits.begin() and the loop exits right away so you are safe.

bechenard replied:
Just checked, it works with a single entry.

brownd1978 commented on CaloCluster/src/CaloClusterFast_module.cc:
```diff
@@ -1,322 +1,197 @@
+              std::vector<int>  neighborsId = cal.crystal(visitId).neighbors();
+              if (extendSearch_) std::copy(cal.nextNeighbors(visitId).begin(), cal.nextNeighbors(visitId).end(), std::back_inserter(neighborsId));
+
+              for (auto& iId : neighborsId)
```
I think this could be auto const&

bechenard replied:
Almost, this should obviously be const auto& !!!!!!

brownd1978 commented on CaloCluster/src/CaloClusterFast_module.cc:
```diff
@@ -1,322 +1,197 @@
+      
+      auto functorTime = [&caloHits](int a, int b) {return caloHits[a].time() < caloHits[b].time();};
+      std::sort(hits.begin(),hits.end(),functorTime);
+      for (auto val : hits) if (caloHits[val].energyDep() < EnoiseCut_) val = -1;
```
could you make this cut 1 step earlier, ie only fill the hits index array with hits above thershold.  This would reduce the number of downstream operations.

brownd1978 replied:
That would also eliminate the clumsy logic having to interpret -1 below.

bechenard replied:
Yes. In practice, almost all hits will pass the cut but I have done the change anyway

brownd1978 commented on CaloCluster/src/CaloClusterFast_module.cc:
```diff
@@ -1,322 +1,197 @@
+
+          //start the clustering algorithm for the hits between iStart and iStop
+          std::queue<int> crystalToVisit;
+          std::vector<bool> isVisited(cal.nCrystal()); 
```
I don't understand why the isVisited array needs to be so large .  Couldn't you make a struct (int, bool) to combine this with the hits index?  That would both reduce the size and keep the related data close in memory.

bechenard replied:
As I say below, you need to keep track independently of visited crystal ID during a clustering iteration, and hits being used (see below). Clustering typically visits 50-100 crystal IDs with the double ring algorithm (but many IDs are seen several times, so if you include them all regardless of the uniqueness, you have several hundred entries), and you need a "not so small structure" to keep track of them - unless you use a set which is slow. On the other hand, a vector of bool is packed (1 bit / value instead of 8 bits), so you actually need 1800 bits to keep track of all crystals, which isn't that different from O(100) ints. I did some tests and this is the fastest solution I found so far. You might gain O(50 us) is you really optimize like crazy, but I prefer to keep the code simple here.

brownd1978 commented on CaloCluster/src/CaloClusterFast_module.cc:
```diff
@@ -1,322 +1,197 @@
+
+                      if (caloHits[*it].energyDep() > ExpandCut_) crystalToVisit.push(iId);
+                      clusterList.push_back(*it);
+                      *it = -1;
```
using -1 as a flag is a bit clumsy.  If you combined index and isVisited you could get rid of it.

bechenard replied:
There is a subtlety here. You need to keep track of two things: 1) the crystals you visit when you perform the clustering algorithm - otherwise you keep looping over the same neighbors forever - and 2) the hits that have been included in a cluster. Note that the isVisited structure needs to be cleared every time the clustering algorithm - a crystal might have two hits at different times and might need to be inspected several times. So the two structures serve different purposes and I don;t think they should be combined.

bechenard replied:
Since you already have a structure that holds the hit indices, the most economical thing to do is reuse it and simply use a special flag to mark a crystal already used.

brownd1978 reviewed (commented):
Just a few comments about clustering now I found the module.

FNALbuild commented:
:memo: The latest commit by @echenard is timestamped 0:04:53.930228 (hh:mm:ss) in the future. Please check that the date and time is set correctly when creating new commits.

brownd1978 reviewed (approved):
thanks for going through the comments.

gianipez commented:
Hi @bechenard , I think you still need to process the last set of comments I posted: https://github.com/Mu2e/Offline/pull/398#pullrequestreview-631737107
Without it, it will be impossible to run the trigger paths in the Offline.

Cheers,

Giani

bechenard commented:
@gianipez I added filters and sequences in the trigger prolog in CaloFilter, and make the change you requested in the filter module. If this is not sufficient, then it might be more efficient if you take care of this (or tell me exactly what needs to be changed / added). If the code works with the current way of making CaloDigis, then I suggest we merge it and make another PR to simulate the online CaloDigiMaker (it might not be 100% compatible with the current offline workflow as hits are extracted and merged in different ways even with the RawProcessor)

gianipez commented:
if the nightly doesn't run the trigger that's fine with me. If the nightly test runs the trigger paths, then it will crash.

brownd1978 commented:
Currently the nightly does not and cannot run the trigger.  Once we have
migrated to fixed paths it will be possible, though we may have to
introduce a new validation job for that.

Dave

On Tue, Apr 13, 2021 at 12:17 PM Gianantonio Pezzullo <
***@***.***> wrote:

> if the nightly doesn't run the trigger that's fine with me. If the nightly
> test runs the trigger paths, then it will crash.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/398#issuecomment-818990543>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57ZBGSZQYRUVE62CF63TISKEHANCNFSM42GFP4LA>
> .
>


-- 
David Nathan Brown            ***@***.***
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gianipez reviewed (approved):
ok then. I'm currently working more on the online side these days, so I'm not sure when I'll be capable to look at the offline CaloHit sequence

bechenard commented:
So I would propose we go ahead with it and Giani or somebody else makes the required changes to run the trigger in another PR.

gianipez reviewed (approved):
that's fine with me (forgot to approve all the files before)

gianipez reviewed (approved):
ok

gianipez commented:
please jenkins try a build

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for d5d9fa715801eaefe6f1858e860c10373ae5e1dc: build  (Build queue is empty)

FNALbuild commented:
:sunny: The tests passed at d5d9fa715801eaefe6f1858e860c10373ae5e1dc.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :white_check_mark: | Merged d5d9fa715801eaefe6f1858e860c10373ae5e1dc at e2057a9a842cac06fa5dac12d32e29f2f58f38ae |
| scons build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/scons.log). Build time: 13 min 59 sec |
| ceSimReco | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/ceSimReco.log) |
| g4test_03MT | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/g4test_03MT.log) |
| transportOnly | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/transportOnly.log) |
| PS | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/PS.log) |
| g4study | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/g4study.log) |
| cosmicSimReco | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/cosmicSimReco.log) |
| rootOverlaps | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/rootOverlaps.log) |
| g4surfaceCheck | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :white_check_mark: | [TODO (0) FIXME (0) in 10 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [3 errors 4 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//497/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

FNALbuild commented:
:memo: The HEAD of `master` has changed to ab9ee5498f5d3d8fbd15582e928bb9685bd20e53. Tests are now out of date.

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for d5d9fa715801eaefe6f1858e860c10373ae5e1dc: build  (Build queue has 3 jobs)

FNALbuild commented:
:sunny: The tests passed at d5d9fa715801eaefe6f1858e860c10373ae5e1dc.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :white_check_mark: | Merged d5d9fa715801eaefe6f1858e860c10373ae5e1dc at ab9ee5498f5d3d8fbd15582e928bb9685bd20e53 |
| scons build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/scons.log). Build time: 14 min 02 sec |
| ceSimReco | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/ceSimReco.log) |
| g4test_03MT | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/g4test_03MT.log) |
| transportOnly | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/transportOnly.log) |
| PS | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/PS.log) |
| g4study | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/g4study.log) |
| cosmicSimReco | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/cosmicSimReco.log) |
| rootOverlaps | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/rootOverlaps.log) |
| g4surfaceCheck | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/g4surfaceCheck.log) |
| FIXME, TODO count | :white_check_mark: | [TODO (0) FIXME (0) in 10 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [3 errors 4 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/artifact/clang-tidy.log) |


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//499/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.
